crypt.go

```diff
@@ -193,20 +207,64 @@ func (s *activateWithKeyDataState) run() (success bool) {
 			continue
 		}
 
-		return true
+		return true, nil
 	}
 
-	// TODO: Passphrase support
+	// Try keys that require a passphrase
+	tries := s.passphraseTries
+	var lastErr error
+
+	for tries > 0 && numPassphraseKeys > 0 {
+		tries -= 1
+
+		// Request a passphrase first and then try each key with it. One downside of
+		// this approach is that if there are multiple keys with different passphrases
+		// or the passphrase is wrong for all keys, this will accelerate the rate at
+		// which dictionary attack protections kick in for platforms that support that.
+		// This shouldn't be an issue for standard configurations where there would be
+		// a maximum of 2 keys with passphrases enabled (Ubuntu Core based desktop on
+		// a UEFI+TPM platform with run+recovery and recovery-only protectors for
+		// ubuntu-data).
+		passphrase, err := s.authRequestor.RequestPassphrase(s.volumeName, s.sourceDevicePath)
+		if err != nil {
+			lastErr = xerrors.Errorf("cannot obtain passphrase: %w", err)
+			continue
+		}
+
+		for _, k := range s.keys {
+			if k.AuthMode()&AuthModePassphrase == 0 {
+				continue
```

Comment:
tests don't seem to reach here, so there's not test for mixing different modes?

Reply:
I've added a test for this.